### ISSUE TITLE
Invalid pdf error message for form detail landing page

### ISSUE THREAD

**Issue by jenniferlee-dsva** [department-of-veterans-affairs/va.gov-team#12181]
## Invalid pdf error message for form detail landing page

_Description: We need an error message for invalid PDF links._ 

**Use case/steps:** 

1. A user is on a form detail page. 
2. They click on the 'Download' form PDF link. 
3. The PDF link is invalid. 
4. Today in the legacy form detail page experience when a PDF link is invalid, the user is taken to our generic Not Found 404 page. 

### Background
Because this situation happens more frequently than it should and particularly with forms, we want to message the user with something more custom and advise them to contact the VA forms manager who will be able to resolve the issue at the source VA forms DB, where the the PDF links are actually generated. 

[Background of use case - Slack thread](https://dsva.slack.com/archives/CUB5X5MGF/p1596123199101700)

Also, there are some forms with PDF with the `/vaforms/` path in their urls. When we sunset the legacy VA Forms experience, I imagine these urls will become invalid. At that point, it'll fall under this issue's use-case. `MS 8/27`

### New message: [ updated 8/17 per [comment](https://github.com/department-of-veterans-affairs/va.gov-team/issues/12181#issuecomment-674948198)]

 **This form link isn't working**
We're sorry, but the form you tried to download appears to have an invalid link. Please [email the VA forms managers](VaFormsManagers@va.gov) for help with this form. 

**Note:** 
Please consult ux designer @andaleliz about the recommended design of this experience. 

- Should the message display in an alert box on the form detail page?  - yes
- Should the user be taken off the page and sent to a custom content version of the 404? - no

## Tasks
- [x] decide ux/design
- [ ] implement on detail pages

## Acceptance criteria
- [x] Error message implemented for invalid PDF scenario on form detail pages
- [x] Alert should appear AFTER the user clicks on the link - just like a regular error message.
- [x] Okay to hide the bad link with the message after that point, but include  the bad url in the message of the email for the user:  `I tried to download this form but the link doesn't work: [ insert bad link ] `
- [x]  Provide a recording of this use case for @tzelei (showing use case steps generating this error message including mailto email).  {Added 8/19} -- instead of a recording, we provided a full staging environment dummy.

**Comment by ncksllvn** [comment]
The form detail pages will be static HTML generated during the build from Drupal data (like other Drupal pages) and right now that Drupal data does not include any indicator of whether a form's PDF leads to a 404 or not. I think we have a couple options then -

1. As CMS work, generate a new field during each migration called `fieldFormPdfValid` (or similar) that indicates whether a form's download URL is valid or not. (a value of false = 404, indicating invalid)
1. From the browser, on page load of each form detail page, send an HTTP request to determine whether a form's URL leads to a 404.
    - I was just messing with this. I was thinking browser security would block us from checking the status code of forms on `www.vba.va.gov`, but it looks like we still can. @kelsonic from www.va.gov/find-forms/ try running `fetch('https://www.vba.va.gov/pubs/forms/VBA-21-0819-ARE.pdf', { mode: 'no-cors' });` (should be a 200) and then another doing `fetch('https://www.vba.va.gov/pubs/forms/VBA-21-0819NONEXISTENT-ARE.pdf', { mode: 'no-cors' });` (should be a 404)

**Comment by ncksllvn** [comment]
I'm (cautiously) going to say that I think we are okay to get started on this via the second option. I don't think we have any dependencies but we may find limitations with this approach going forward have to do option 1 (the CMS approach)

**Comment by kelsonic** [comment]
Cautiously fail fast (or succeed fast if we're lucky!), love it! ❤️

**Comment by andaleliz** [comment]
Couple of background questions:
1. How do these links get put in to begin with? 
2. Are they managed by business-side content authors outside of our VSA team?
3. I'm assuming the links go bad because someone deletes/moves a file and doesn't update something... Is that right?

Second, what's the timing of this? Looks like it's not associated w/ a milestone but also doesn't have `needs-grooming`.

**Comment by jenniferlee-dsva** [comment]
> How do these links get put in to begin with?

They're generated by the file name of the form + other form data (e.g.,  administration, etc.) inside the VA forms database.


> Are they managed by business-side content authors outside of our VSA team?

Business side - VA forms managers for each administration


> I'm assuming the links go bad because someone deletes/moves a file and doesn't update something... Is that right?

Yes

> Second, what's the timing of this? Looks like it's not associated w/ a milestone but also doesn't have needs-grooming.

I'm going to defer to @MickinSahni on this. We do want this ready for Tze's QA review. I don't think we have an eta date for QA yet. Here is our QA ticket for Tze - #11335

**Comment by jenniferlee-dsva** [comment]
Clarification ^ - we don't have an ETA for QA date yet, but I think they're waiting on us to let them know what the product experience is set to QA with. 

So - correct me if I'm off @MickinSahni - we should complete this task to unblock Tze's QA ticket?

**Comment by MickinSahni** [comment]
> Clarification ^ - we don't have an ETA for QA date yet, but I think they're waiting on us to let them know what the product experience is set to QA with.
> 
> So - correct me if I'm off @MickinSahni - we should complete this task to unblock Tze's QA ticket?

that's generally correct, @jenniferlee-dsva. but checking in with the QA team on when they anticipate being able to prioritize completing test cases they already started stubbing out (https://github.com/department-of-veterans-affairs/va.gov-team/issues/11335#issuecomment-671143420)

**Comment by ncksllvn** [comment]
I think this one is required for MVP, so I think we should consider get this one ready for implementation. Before that, we need two things -

1. Where should the error be placed, and what should it look like. 
    - My gut says top of the page as a [warning alertbox](https://design.va.gov/components/alertboxes#warning-alert)
1. The content of the alert box

Engineering work _can_ get started and then adjust the error once finalized, but I'd rather  we have that upfront.

**Comment by jenniferlee-dsva** [comment]
Content of the message can say: 

**New message:**
We're sorry, but the form you're trying to download appears to have an invalid link. Please contact the VA forms managers for more information about this form.

I'm good with alert box on the page, but going to defer here to Liz on exact location of the alert box. (Only requirement is that it's an in-body alert box, not a full-width banner that goes across the top of the whole page.)

**Comment by jenniferlee-dsva** [comment]
Once this is complete, can we capture it as a screenshot for @tlei123's QA test?

**Comment by ncksllvn** [comment]
I think so, as long as we have the URL of a form that meets this criteria

**Comment by jenniferlee-dsva** [comment]
Hi @ncksllvn 
We fixed a recent example of this scenario. There may be others that we don't know about - bc form managers are changing/updating forms all the time in the source DB. 

If I gave you that previous known example of an invalid form pdf URL, could you use it to test but without us having to revert versions in prod to that bad form version?  (This kind of got us into problems.) 

On this node http://preview-prod.vfs.va.gov/preview?nodeId=5776 for VA Form 10-0137:

- Currently it is showing the CORRECT and valid form pdf link: https://www.va.gov/vaforms/medical/pdf/VA Form 10-0137 FILL.pdf

- But in an earlier revision history, it had an invalid form pdf link: https://www.va.gov/vaforms/medical/pdf/VA Form 10-0137(1).pdf  [per this Slack](https://www.va.gov/vaforms/medical/pdf/10-10EZ-fillable.pdf?_ga=2.193048877.1893908243.1597071845-724846401.1559850749)

Let me know if you can test with this info in staging (hopefully) without the bad form flowing up to prod.

**Comment by jenniferlee-dsva** [comment]
In case the version history helps: 

![image.png](https://images.zenhubusercontent.com/5bb2975c143b4d2ee7a35ef6/82bc2100-c9cf-4cd4-8c71-5d7448453426)

**Comment by andaleliz** [comment]
How can I see an example of a form detail page? Here's what I've tried
- clicking links in the comments above (get a 404).  
- gone to va.gov/find-forms and searched for form numbers (1995, 1010ez 10-10ez), and keywords (pension, disability), but all the results give me a download link directly in the search results that points away from VA.gov. 

A couple of thoughts/questions in the meantime:

What happens when a user clicks a broken PDF from the search results themselves? Do we need UX recommendations for that?  We may need a separate design-specific ticket to cover that work.

Is there another way a user can contact the VA forms managers? I'm not positive, but having a text linked to an email like that may be problematic for folks who don't have an email client set up.  At a minimum, we could set a better expectation of what would happen upon clicking that link if we changed it to `email` the VA forms managers, instead of `contact`.

**Comment by jenniferlee-dsva** [comment]
Hi @andaleliz - You need to be on your VA network in order to view the CMS preview pages. (You don't however need CMS access.)

You can also see sample mockups here: https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/find-va-forms-redesign/form-detail-landing-pages/final-docs#visual-mockups 

Note that the content fields are dynamic:

- Some form details have the When to use this form content, some don't
- Some form details have a tool button, some don't
- Some form details have Related forms and instructions, some don't
- Etc.

 

Whereever the invalid form link message appears, please make sure that it will work well across those dynamic variations.

**Comment by ncksllvn** [comment]
>How can I see an example of a form detail page?

Yeah what Jen said - VA network or the SOCKS proxy. Here's a [screenshot of the form detail page for 10-10EZ](https://user-images.githubusercontent.com/1915775/90022701-d83ca880-dc80-11ea-9bb4-9eb8da216686.png) though, which the preview link should have taken you to. Ignore the "you are viewing a draft" message and "Edit this page in the CMS" link.

> What happens when a user clicks a broken PDF from the search results themselves? Do we need UX recommendations for that? We may need a separate design-specific ticket to cover that work.

I think that is out of scope for this

**Comment by jenniferlee-dsva** [comment]
>What happens when a user clicks a broken PDF from the search results themselves?

This alert message for the form detail page only. 

In the search if a form has an invalid PDF, the API/FE logic suppresses the form from appearing at all in search results. I have a separate request out to the LH team to send form managers an autonotification of forms with invalid PDFs, so the form managers can resolve them at the source.

**Comment by ncksllvn** [comment]
@jenniferlee-dsva I missed your comment, sorry about that -

> Once this is complete, can we capture it as a screenshot for @tlei123's QA test?

I can definitely capture a screenshot and locally I can test using just a spoofed form URL to trigger the message. During formal QA though, Tze may need a URL that is set up to trigger the invalid-pdf message. I think we have what we need for now though 🙂

**Comment by andaleliz** [comment]
Thanks, and sounds good. I'll give this my full attention after synthesis is published for LC MVP (should be by EOD today)

**Comment by andaleliz** [comment]
Based on all the possible variations, I think the best place for this is to replace the PDF link. We don't want to send people away since there may be additional helpful information on page, or an online version a person could complete. 

I put together two mock-ups with [minimum](https://vsateams.invisionapp.com/share/G2YF5AJC845#/428497812_detail_Page_With_Minimum_Components_-_Error) and [maximum](https://vsateams.invisionapp.com/share/G2YF5AJC845#/428497810_detail_Page_With_All_Components_-_With_Error) components.

At this point, we _think_ it's technically feasible to detect if the link is invalid before the user clicks. If that's the case, we'll just show the alert on page load.  If it's not, [the full mock-up](https://vsateams.invisionapp.com/share/G2YF5AJC845) shows the experience a person would have if they clicked on an invalid link.

I reviewed with @ncksllvn via Slack and he's good to go - @jenniferlee-dsva let me know if you have any questions or concerns. Thanks!

**Comment by andaleliz** [comment]
Hold up! I just realized this is really closely related to #11380. Looking at this again to take the 2 alerts on one page into consideration.

**Comment by andaleliz** [comment]
Alright, after exploring our options, I still think this is the way to go. I added some screens ([min](https://vsateams.invisionapp.com/share/D8YF92DRX3V) and [max](https://vsateams.invisionapp.com/share/SPYF92NYMBZ) components) to the prototype where you can see the invalid PDF on pages that also have an info alert. 

Not ideal, but I think it clearly communicates information.  My understanding is that this would also be a rare use case to have both on the same page.

**Comment by jenniferlee-dsva** [comment]
@andaleliz - This is so helpful! Could we swap the order of how these would appear if they both appeared (yes, rare)? 

- It would be nice if the informational alert - if used - always appeared below the tool CTA button (if there's one). That way, informational content isn't given more priority over the tool CTA or the form download. 

- RE UX: Want to confirm I understand the flow: the error message appears AFTER someone clicks on the broken form link? I think it could be helpful for the user and form owner to include the bad PDF URL in the email for the form manager, so is there a way we can leave the form link visible so the user can copy/paste it into their email?

**Comment by jenniferlee-dsva** [comment]
Header text feedback - Instead of "This form is not available," let's use: `This form link isn't working`

We want to avoid saying 'not available' bc that could be misinterpreted to mean that the form is removed or retired by VA, which isn't the case with these invalid PDF links.

**Comment by andaleliz** [comment]
@jenniferlee-dsva you're more familiar with the variety of information alerts that could appear on these pages. If none of that information would ever be more important than the CTA, it definitely makes sense to put it below. No prob to make that update.

>I think it could be helpful for the user and form owner to include the bad PDF URL in the email for the form manager, so is there a way we can leave the form link visible so the user can copy/paste it into their email?

I see your point that it _could_ be helpful. I'm hesitant to recommend displaying it for a few reasons:
-  People don't always read alerts. They could skip over it and click the link anyway which could cause confusion and frustration.
- As a person using a website, if I came across a bad link with a notice saying the link was bad, I would be annoyed and ask myself "why wouldn't they just take it down or fix it if they know it's bad"? I wouldn't want to make anyone visiting our site feel that way.
- It's really not on the user to help the form manager. They just care about getting the form. My understanding of the alert's intention for the Veteran is to give them a way to get the form, not to report it.

It sounded to me from the slack thread that a team (Charley and Michael?) are working on auto-notifications to alert the form managers when a bad PDF URL is detected.  If that's the case, won't they know about it anyway?

**Comment by andaleliz** [comment]
I have some more thoughts on the text in the alert box:

- Instead of `This form link isn't working` can we say something like `The form is temporarily unavailable for download`? Users won't necessarily know what form link we're talking about, or even expect a link, when they land on the page, especially if we hide it.

- Can we expose the email address in the text so that someone who doesn't have an email client set up on their device can still know what email address they should send it to? E.g. I don't have email set up on my phone, so if I saw that on mobile, I wouldn't be able to do anything as it is. But if I saw the email address, I could write it/save it, and send the email later when I have that available or give it to someone who can send an email on my behalf.  

- I'm assuming we're having people send an email to give them a chance to say "hey, since I can't download this, please email me the form".  If that is the case, can we make it more clear that that's an option? Also, I think the "you're trying to download" part isn't exactly correct, because the user could see this on the page without actually trying to do anything (e.g. they go to a detail page w/ download and online options, and they just want to fill it out online)

**We're sorry, but this form appears to have an invalid download link. Please email [vaformmanagers@va.gov](vaformmanagers@va.gov) to request the PDF.**

If forms managers will get notifications when this is detected, and can commit to fixing within a number of business days, we could even offer an alternate next step:
**We're sorry, but this form appears to have an invalid download link. Please email [vaformmanagers@va.gov](vaformmanagers@va.gov) to request the PDF or check back in 2 business days.**

**Comment by jenniferlee-dsva** [comment]
@andaleliz  - bummer - I was trying to answer in-line ^ and Github lost all my replies! So instead I will summarize here but kind of keeping it short bc I don't want to recreate all the reasons. 

**Summary:** 

- Alert should appear AFTER the user clicks on the link - just like a regular  error message. 
- Okay to hide the bad link with the message after that point, but then could we plop the bad url into the message of the email for the user? 
> Example: "I tried to download this form but the link doesn't work: [ insert bad link ] " 

**Updated message header and text updates based on ^ user flow:** 

> **This form link isn't working** 
> We're sorry, but the form you're tried to download appears to have an invalid link. Please email the VA forms managers for help with this form. [ can expose email address if it won't cause bot spam ] 

Concerned about exposing email address because I thought that made it easy for bots to scrape email addresses for spamming.

Also re "temporarily unavailable for download" - To me, this sounds vaguely like it could be a system issue. I want it to be very clear to users and to form managers that it is the PDF form link itself that is the problem -- not anything to do with the site, the cms, the internet, etc.

**Comment by andaleliz** [comment]
The alert copy updates make sense based on the flow you outlined - thank you!

>could we plop the bad url into the message of the email for the user?

I think that's a good solution.

>Concerned about exposing email address because I thought that made it easy for bots to scrape email addresses for spamming.

Nick can speak to that better than I can. If it does cause spam, is there a contact form or something else we can offer to people? I have personally felt the frustration of these kinds of links online and not been able to do anything because I couldn't see the email address; I'd hate to do that to anyone else.

**Comment by andaleliz** [comment]
@ncksllvn what are your thoughts on Jen's concerns RE exposing the email address and bot spam?

**Comment by ncksllvn** [comment]
> @ncksllvn what are your thoughts on Jen's concerns RE exposing the email address and bot spam?

I don't know enough about that to say confidently...sometimes email addresses are written as `someone(at)va.gov` so that a person can read it but a bot won't try. 

@kelsonic (no need to read this whole thread) - do you know if publishing an email address onto a site with a lot of traffic will result in a bunch of spam sent to that email via bots?

**Comment by kelsonic** [comment]
Hmm, so don't have a lot of context here but a new email address has _the potential_ to get spam from bots but there's definitely solutions to that -- such as utilizing spam filters on the email client, etc.! Not sure if that helps at all 😅

**Comment by jenniferlee-dsva** [comment]
We don't own this email address - it is the shared inbox for the VA central, VBA, NCA, and VHA form managers. This email box isn't something new - they've already been using it on the legacy search experience -- but as a mailto click, not with the email address itself written out. 

So given that we don't have a way to ensure bot scraping for spam, let's keep the existing experience of having the email address hidden behind the link. 

**I'll summarize it here, but will also update the og ticket AC:** 

-  Alert should appear AFTER the user clicks on the link - just like a regular error message.
-  Okay to hide the bad link with the message after that point, but include  the bad url into the message of the email for the user like:  `I tried to download this form but the link doesn't work: [ insert bad link ] `

**Update message header and text updates based on ^ user flow and not showing the email address:**

 **This form link isn't working**
We're sorry, but the form you tried to download appears to have an invalid link. Please [email the VA forms managers](VaFormsManagers@va.gov) for help with this form.

**Comment by andaleliz** [comment]
Prototype has been updated to reflect the above/

**Comment by ncksllvn** [comment]
Ah man, so I have disappointing news...While implementing this, I've run into the same issue we ran into with attempting to download straight to the tray - we can only check the PDF's validity when it's on our domain, www.va.gov. This means that from our detail pages, we won't be able to check the validity of PDFs hosted on `www.vba.va.gov`. Sorry, I unknowingly didn't do a deep enough tech dive in the beginning 😞

**Comment by jenniferlee-dsva** [comment]
Hmm - correct me if I'm wrong, but I don't think we need to do we? I thought that our forms search -- through the LH API? -- currently can tell when a pdf  is not valid. (That's how it is able to supress invalid forms from displaying in results even when people search for it by form number/name.) 

I thought that we would be able to use that same API `valid_pdf` flag to know which form pdf has a bad URL on our form detail pages?

**Comment by ncksllvn** [comment]
Oh, good point. I forgot about that option @jenniferlee-dsva. I'll try that!

**Comment by jenniferlee-dsva** [comment]
@andaleliz - on the 'max' version of mockup, could you include where the informational alert message would appear, so it's clear what the header hierarchy would be? (Below the online tool CTA button)

https://vsateams.invisionapp.com/share/G2YF5AJC845#/screens/428497810_detail_Page_With_All_Components_-_With_Error

**Comment by andaleliz** [comment]
@jenniferlee-dsva Is the next screen in the prototype what you're looking for?
https://vsateams.invisionapp.com/share/G2YF5AJC845#/screens/428525873

**Comment by ncksllvn** [comment]
Do we want this functionality also for PDF links included under the Related Forms sections?

![image.png](https://images.zenhubusercontent.com/59cb9de1b0222d5de47953d8/5a10a244-d19e-4bfb-970f-46588143df87)

I'm assuming we do. It doesn't really add much to the LOE to implement this.

**Comment by jenniferlee-dsva** [comment]
That's great, thank you @andaleliz. (I didn't know there was a next page!) 

@ncksllvn - I just realized something - I'm not sure we even need the API valid_pdf flag for this, because the experience is: 

1. User clicks on the link. 
2. The link 404s. 
3. We display the alert message. 

So wouldn't we be able to tell just from the form link breaking upon the user clicking? Nvmd me though if I am missing something technical - tot possible.

**Comment by jenniferlee-dsva** [comment]
@ncksllvn - re https://github.com/department-of-veterans-affairs/va.gov-team/issues/12181#issuecomment-675726062

Oh, interesting! Yes, great catch. 

@andaleliz - Could you see Nick's comment above? I would imagine that the same error message can display instead of the bad form link -- just like the other one, after clicking.

**Comment by ncksllvn** [comment]
> I just realized something - I'm not sure we even need the API valid_pdf flag for this, because the experience is:
> 
> 1. User clicks on the link.
> 2. The link 404s.
> 3. We display the alert message.

I was thinking that originally actually but we won't be able to tell the status (404) of anything outside of our domain, www.va.gov, even though we linked to the download. So we have to rely on the API to tell us about the PDF's validity. 

I was thinking that we don't do number 2 btw - we would essentially use JS to stop the download from proceeding until we determine the PDF's validity, then only allow it to download if it is valid

**Comment by andaleliz** [comment]
@jenniferlee-dsva yes, I think that should be the behavior for those links too. But, imagine a page with all the links broken and an info alert (unlikely worse-case scenario) - eek! 

It makes me want a lighter weight treatment for the error alert, maybe without a heading.  The design system component always has a heading in the examples, and same with the Sketch symbol. But, the guidance doesn't say that it _must_ have a heading. @ncksllvn what happens if you put an alert in the code without a heading? Does the alert message align with the icon, or is there a gap where the heading should be?  

For our immediate purposes here and now, I think we can just move forward with the alert as-is, headings help with scanability.  This is something to consider iterating on in the future though.

**Comment by jenniferlee-dsva** [comment]
@andaleliz - for this ticket, agree: let's use the same alert as-is. 
RE headings - that sounds like a thing for the design system council - Ryan, Martha, Kevin, Emily - to consider for future and global pattern guidance. Good points though. 

@ncksllvn - Sorry to add one more AC, but bc we're not going to recreate this scenario in staging for QA, please provide a recording for Tze. He asked for a screenshot, but I think we need a recording here like we did for sitewide system downtime messages because there are some steps and behaviors here that a screenshot won't be able to convey. 

>  [ ]  Provide a recording of this use case for @tzelei (showing use case steps generating this error message including mailto email).  {Added 8/19}

fysa  @MickinSahni

**Comment by ncksllvn** [comment]
This work will go live today. I also added a static HTML page on GItHub to help us see the UX. The page will be published to https://staging.va.gov/form-detail-dummy and should make it there before EOD.

**Comment by jenniferlee-dsva** [comment]
@ncksllvn - thx for the test dummy page. the only part of the AC that's missing is autopopulating the bad form URL in the mailto email message text. 

So when I click on the form manager email link and the email message client opens, it should populate the To: with the form manager email (which it does now) + in the message body, it should autopopulate the text: 

> I tried to download this form but the link doesn't work: [ autopopulate the bad link ]

**Comment by ncksllvn** [comment]
Oh interesting, I'll see if I can add that!

**Comment by MickinSahni** [comment]
Nick working on the email update and an example for of the invalid PDF url for the product guide and QA

**Comment by ncksllvn** [comment]
Note to self - Make this page a complete form detail but with invalid PDF links - https://staging.va.gov/form-detail-dummy - to show full context

**Comment by jenniferlee-dsva** [comment]
@ncksllvn - The staging full context looks great! Could we give @tzelei a recording, so he can validate the invalid PDF experience?   -- OR @tzelei - would you be okay just looking at the staging experience as your validation?  The link is here and it shows thefull context. Just click on the form PDF link; and then click on the 'email form managers' link in the error message. 

https://staging.va.gov/form-detail-dummy

**Comment by tlei123** [comment]
> @ncksllvn - The staging full context looks great! Could we give @tzelei a recording, so he can validate the invalid PDF experience? -- OR @tzelei - would you be okay just looking at the staging experience as your validation? The link is here and it shows thefull context. Just click on the form PDF link; and then click on the 'email form managers' link in the error message.
> 
> https://staging.va.gov/form-detail-dummy

@jenniferlee-dsva , @ncksllvn , I'll check this out this afternoon.  I shouldn't need a recording now, since this is up on Staging.  Thx much!

**Comment by jenniferlee-dsva** [comment]
I believe we can close this ticket, as we have completed each of the AC's here. @tzelei please let us know in the QA ticket re QA status.
